kafka-0.10.x/src/main/scala/monix/kafka/KafkaConsumerObservable.scala

```diff
@@ -62,44 +65,61 @@ trait KafkaConsumerObservable[K, V, Out] extends Observable[Out] {
-      }
-      feedTask.runAsync(cb)
+      val startConsuming =
+        Resource
+          .make(consumer) { c =>
+            // Forced asynchronous boundary
+            Task.evalAsync(consumer.synchronized(blocking(c.close()))).memoizeOnSuccess
+          }
+          .use { c =>
+            // Skipping all available messages on all partitions
+            if (config.observableSeekOnStart.isSeekEnd) c.seekToEnd(Nil.asJavaCollection)
+            else if (config.observableSeekOnStart.isSeekBeginning) c.seekToBeginning(Nil.asJavaCollection)
+            // A task to execute on both cancellation and normal termination
+            pollHeartbeat(c)
+              // If polling fails the error is reported to the subscriber and
+              // wait 1sec as a rule of thumb leaving enough time for the consumer
+              // to recover and reassign partitions
+              .onErrorHandleWith(ex => Task(out.onError(ex)))
+              .loopForever
+              .startAndForget
+              .flatMap(_ => runLoop(c, out))
+          }
+      startConsuming.runAsync(cb)
     }
   }
 
-  /* Returns a task that continuously polls the `KafkaConsumer` for
-   * new messages and feeds the given subscriber.
-   *
-   * Creates an asynchronous boundary on every poll.
-   */
+  /** Returns a task that continuously polls the `KafkaConsumer` for
+    * new messages and feeds the given subscriber.
+    *
+    * Creates an asynchronous boundary on every poll.
+    */
   private def runLoop(consumer: Consumer[K, V], out: Subscriber[Out]): Task[Unit] = {
     ackTask(consumer, out).flatMap {
       case Stop => Task.unit
       case Continue => runLoop(consumer, out)
     }
   }
 
-  /* Returns a `Task` that triggers the closing of the
-   * Kafka Consumer connection.
-   */
-  private def cancelTask(consumer: Consumer[K, V]): Task[Unit] = {
-    // Forced asynchronous boundary
-    val cancelTask = Task.evalAsync {
-      consumer.synchronized(blocking(consumer.close()))
-    }
-
-    // By applying memoization, we are turning this
-    // into an idempotent action, such that we are
-    // guaranteed that consumer.close() happens
-    // at most once
-    cancelTask.memoize
+  /** Returns task that constantly polls the `KafkaConsumer` in case subscriber
+    * is still processing last fed batch.
+    * This allows producer process commit calls and also keeps consumer alive even
+    * with long batch processing.
+    *
+    * @see [[https://cwiki.apache.org/confluence/display/KAFKA/KIP-62%3A+Allow+consumer+to+send+heartbeats+from+a+background+thread]]
+    */
+  private def pollHeartbeat(consumer: Consumer[K, V]): Task[Unit] = {
+    Task.sleep(config.observablePollHeartbeatRate) *>
+      Task.evalAsync(
+        if (!isAcked) {
+          consumer.synchronized {
+            val records = blocking(consumer.poll(0))
+            if (!records.isEmpty) {
+              throw new IllegalStateException(s"Received ${records.count()} unexpected messages.")
```

Comment:
If we raise an error here, a bit more idiomatic would be

```scala
Task.suspend {
 // poll
  Task.raiseError(new IllegalStateException)
}
```

If we don't propagate an error and only report it to the scheduler, `Task.deferAction` will be handy

Reply:
The thing here is that we have to wrap the poll with `consumer.synchronized` which returns `Any`, so that if 
we return a task it would not be evaluated...